pedsnetdcc/tests/prepdb_test.py

```diff
@@ -0,0 +1,239 @@
+        return re.match("(?:" + regex + r")\Z", string, flags=flags)
+    re.fullmatch = fullmatch
+
+
+def introspect_schemas(conn_str):
+    """Return list of schemas in the database
+
+    A future version of this for more generic testing might produce
+    a nested data structure with tables under each schema ....
+
+    :param conn_str: libpq/psycopg2 connection string
+    :type: str
+    :return: list of schemas
+    :rtype: list(str)
+    """
+    schemas = []
+    exclude_schemas = (
+        'pg_catalog',
+        'pg_toast.*',
+        'pg_temp.*',
+        'information_schema'
+    )
+
+    with psycopg2.connect(conn_str) as conn:
+        with conn.cursor() as cursor:
+
+            cursor.execute(
+                'select schema_name from information_schema.schemata')
+            for row in cursor.fetchall():
+                schema = row[0]
+                should_exclude = False
+                for pat in exclude_schemas:
+                    if re.fullmatch(pat, schema):
+                        should_exclude = True
+                        break
+                if not should_exclude:
+                    schemas.append(row[0])
+    conn.close()
+
+    return schemas
+
+
+def exec_ddl(conn_str, sql, no_transaction=False):
+    with psycopg2.connect(conn_str) as conn:
+        if no_transaction:
+            conn.set_isolation_level(0)
+        with conn.cursor() as cursor:
+            cursor.execute(sql)
+    conn.close()
+
+
+class TestPrepareDatabase(unittest.TestCase):
+
+    roles = ('peds_staff', 'harvest_user', 'achilles_user', 'dqa_user',
+             'pcor_et_user', 'loading_user', 'dcc_owner', 'pcornet_sas')
+
+    def create_roles(self):
+        """Create roles so grants will not fail.
+
+        :return: None
+        """
+        with psycopg2.connect(self.conn_str) as conn:
+            with conn.cursor() as cursor:
+                for role in TestPrepareDatabase.roles:
+                    cursor.execute('CREATE ROLE {}'.format(role))
+        conn.close()
+
+    def drop_roles(self):
+        """Drop roles created for tests.
+
+        :return: None
+        """
+        with psycopg2.connect(self.conn_str) as conn:
+            with conn.cursor() as cursor:
+                for role in TestPrepareDatabase.roles:
+                    cursor.execute('DROP ROLE {}'.format(role))
+        conn.close()
+
+    def database_exists(self):
+        """Test database existence
+
+        :return: whether the database for the data_model exists
+        :rtype: bool
+        """
+        new_conn_str = _conn_str_with_database(self.conn_str, self.dbname)
+        try:
+            with psycopg2.connect(new_conn_str) as conn:
+                pass
+        except psycopg2.OperationalError as e:
+            if 'does not exist' in str(e):
+                return False
+            else:
+                raise
+        conn.close()
+        return True
+
+    def setUp(self):
```

Comment:
Probably worth changing and simplifying a lot of this using the `testing.postgresql` package.